src/parser/jobs/blm/Leylines.js

```diff
@@ -0,0 +1,100 @@
+import React from 'react'
+import ACTIONS from 'data/ACTIONS'
+import STATUSES from 'data/STATUSES'
+import {Rule, Requirement} from 'parser/core/modules/Checklist'
+import Module from 'parser/core/Module'
+import {ActionLink} from 'components/ui/DbLink'
+import {i18nMark, Trans} from '@lingui/react'
+
+const LEYLINE_DURATION = 30000
+
+export default class Leylines extends Module {
+	static handle = 'leylines'
+	static i18n_id = i18nMark('blm.leylines.title')
+	static title = 'Ley Lines'
+
+	static dependencies = [
+		'checklist',
+	]
+	_CircleOfPowers = {
+		current: null,
+		history: [],
+	}
+	_LeyLineHistory = []
+
+	constructor(...args) {
+		super(...args)
+		this.addHook('cast', {
+			by: 'player',
+			abilityId: ACTIONS.LEY_LINES.id,
+		}, this._onCastLeyLines)
+		this.addHook('applybuff', {
+			by: 'player',
+			abilityId: STATUSES.CIRCLE_OF_POWER.id,
+		}, this._onApplyCircleOfPower)
+		this.addHook('removebuff', {
+			by: 'player',
+			abilityId: STATUSES.CIRCLE_OF_POWER.id,
+		}, this._onRemoveCircleOfPower)
+		this.addHook('complete', this._onComplete)
+	}
+
+	_onCastLeyLines(event) {
+		this.ll = this._LeyLineHistory
+		this.ll.push(event.timestamp)
+
+	}
+
+	_onApplyCircleOfPower(event) {
+		this.cop = this._CircleOfPowers
+		if (this.cop.current) { this.cop.history.push(this.cop.current) }
+
+		this.cop.current = {
+			start: event.timestamp,
+			stop: null,
+		}
+		this._CircleOfPowers = this.cop
+	}
+
+	_onRemoveCircleOfPower(event) {
+		this.cop = this._CircleOfPowers
+		if (this.cop.current) {
+			this.cop.current.stop = event.timestamp
+		}
+		this._CircleOfPowers = this.cop
+	}
+
+	_percentFunction(numberOfLeyLines, SumOfCoPUpime) {
+		return (SumOfCoPUpime/(numberOfLeyLines*LEYLINE_DURATION))*100
+	}
+
+	_onComplete(event) {
+		this.cop = this._CircleOfPowers
+		if (!this.cop.history.length) { return }
+		if (this.cop.current) {
+			if (!this.cop.current.stop) {
```

Comment:
This one's a non-blocker, but - if I'm understanding it correctly, I think you could invert this snippet of logic here so that it only pushes to history if the current event _does_ have a stop time, which should remove the need for the `history.filter()` line below, since you wouldn't be pushing CoP events that hit the end of the fight in the first place.